File: sdk/java-sdk/src/main/java/com/evernym/verity/sdk/protocols/connecting/v1_0/ConnectionsV1_0.java
package com.evernym.verity.sdk.protocols.connecting.v1_0;

import com.evernym.verity.sdk.exceptions.VerityException;
import com.evernym.verity.sdk.protocols.MessageFamily;
import com.evernym.verity.sdk.protocols.Protocol;
import com.evernym.verity.sdk.utils.Context;
import com.evernym.verity.sdk.utils.Util;
import org.json.JSONObject;

import java.io.IOException;

/**
 * An interface for controlling a 1.0 Connections protocol.
 *
 * @see <a href="https://github.com/hyperledger/aries-rfcs/tree/9b0aaa39df7e8bd434126c4b33c097aae78d65bf/features/0160-connection-protocol" target="_blank" rel="noopener noreferrer">Aries 0160: Connection Protocol</a>
 */
public interface ConnectionsV1_0 extends Protocol {
    /**
     * The qualifier for the message family. Uses the community qualifier.
     */
    String QUALIFIER = Util.COMMUNITY_MSG_QUALIFIER;
    /**
     * The name for the message family.
     */
    String FAMILY = "connections";
    /**
     * The version for the message family.
     */
    String VERSION = "1.0";

    /**
     * @see MessageFamily#qualifier()
     */
    default String qualifier() {return QUALIFIER;}
    /**
     * @see MessageFamily#family()
     */
    default String family() { return FAMILY;}
    /**
     * @see MessageFamily#version()
     */
    default String version() {return VERSION;}

    /**
     * Sends the get status message to the connection
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void status(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #status
     */
    JSONObject statusMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     * 
     * @see #status
     */
    byte[] statusMsgPacked(Context context) throws VerityException;

    /**
     * Accepts connection defined by the given invitation
     * @param context an instance of the Context object initialized to a verity-application agent
     * @throws IOException when the HTTP library fails to post to the agency endpoint
     * @throws VerityException when wallet operations fails or given invalid context
     */
    void accept(Context context) throws IOException, VerityException;

    /**
     * Creates the control message without packaging and sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the constructed message (JSON object)
     * @throws VerityException when given invalid context
     *
     * @see #accept
     */
    JSONObject acceptMsg(Context context) throws VerityException;

    /**
     * Creates and packages message without sending it.
     * @param context an instance of the Context object initialized to a verity-application agent
     * @return the byte array ready for transport
     * @throws VerityException when wallet operations fails or given invalid context
     *
     * @see #accept
     */
    byte[] acceptMsgPacked(Context context) throws VerityException;

}

File: sdk/dotnet-sdk/src/VeritySDK/Protocols/Conecting/v1_0/ConnectionsV1_0.cs
using System.Json;
using VeritySDK.Utils;

namespace VeritySDK.Protocols.Conecting
{
    /// <summary>
    /// A class for controlling a 1.0 Connections protocol.
    /// </summary>
    public class ConnectionsV1_0 : AbstractProtocol
    {
        #region Protocol identificator

        /// <summary>
        /// The name for the message family.
        /// </summary>
        public override string qualifier() { return Util.COMMUNITY_MSG_QUALIFIER; }

        /// <summary>
        /// The qualifier for the message family. Uses the community qualifier.
        /// </summary>
        public override string family() { return "connections"; }

        /// <summary>
        /// The version for the message family.
        /// </summary>
        public override string version() { return "1.0"; }


        #endregion 

        #region Constructors

        /// <summary>
        /// Constructor
        /// </summary>
        public ConnectionsV1_0() { }

        /// <summary>
        /// Constructor
        /// </summary>
        public ConnectionsV1_0(string threadId) : base(threadId) { }

        /// <summary>
        /// Constructor
        /// </summary>
        public ConnectionsV1_0(string forRelationship, string threadId) : base(threadId)
        {
            this.forRelationship = forRelationship;
            this.label = null;
            this.base64InviteURL = null;
        }

        /// <summary>
        /// Constructor
        /// </summary>
        public ConnectionsV1_0(string forRelationship, string label, string base64InviteURL)
        {
            this.forRelationship = forRelationship;
            this.label = label;
            this.base64InviteURL = base64InviteURL;
        }

        #endregion 

        string forRelationship;
        string label;
        string base64InviteURL;

        string ACCEPT_INVITE = "accept";
        string STATUS = "status";

        /// <summary>
        /// Sends the get status message to the connection
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public void status(Context context)
        {
            send(context, statusMsg(context));
        }

        /// <summary>
        /// Creates the control message without packaging and sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public JsonObject statusMsg(Context context)
        {
            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(STATUS));
            msg.Add("@id", getNewId());
            msg.Add("~for_relationship", forRelationship);
            addThread(msg);

            return msg;
        }

        /// <summary>
        /// Creates and packages message without sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] statusMsgPacked(Context context)
        {
            return packMsg(context, statusMsg(context));
        }

        /// <summary>
        /// Accepts connection defined by the given invitation 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        public JsonObject acceptMsg(Context context)
        {
            JsonObject msg = new JsonObject();
            msg.Add("@type", messageType(ACCEPT_INVITE));
            msg.Add("@id", getNewId());
            msg.Add("~for_relationship", forRelationship);
            addThread(msg);

            msg.Add("label", label);
            msg.Add("invite_url", base64InviteURL);

            return msg;
        }

        /// <summary>
        /// Creates the control message without packaging and sending it. 
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the constructed message (JSON object)</returns>
        public void accept(Context context)
        {
            send(context, acceptMsg(context));
        }

        /// <summary>
        /// Creates and packages message without sending it.
        /// </summary>
        /// <param name="context">an instance of the Context object initialized to a verity-application agent</param>
        /// <returns>the byte array ready for transport</returns>
        public byte[] acceptMsgPacked(Context context)
        {
            return packMsg(context, acceptMsg(context));
        }
    }
}